src/Lightning/NoiseProtocol/NoiseProtocol.cs

```diff
@@ -0,0 +1,240 @@
+using System;
+using System.Buffers;
+using Microsoft.Extensions.Logging;
+
+namespace NoiseProtocol
+{
+   public class NoiseProtocol : INoiseProtocol
+   {
+      private readonly ILogger<NoiseProtocol> _logger;
+      
+      readonly IEllipticCurveActions _curveActions;
+      readonly IHkdf _hkdf;
+      readonly ICipherFunction _aeadConstruction;
+      readonly IHashFunction _hasher;
+      readonly INoiseMessageTransformer _messageTransformer;
+      readonly IKeyGenerator _keyGenerator;
+
+      public HandshakeContext HandshakeContext { get; set; }
+
+      public NoiseProtocol(IEllipticCurveActions curveActions, IHkdf hkdf, 
+         ICipherFunction aeadConstruction, IKeyGenerator keyGenerator, IHashFunction hasher,
+         INoiseMessageTransformer messageTransformer, byte[] privateKey, ILogger<NoiseProtocol> logger)
+      {
+         _curveActions = curveActions;
+         _hkdf = hkdf;
+         _aeadConstruction = aeadConstruction;
+         _keyGenerator = keyGenerator;
+         _hasher = hasher;
+         _messageTransformer = messageTransformer;
+         _logger = logger;
+         HandshakeContext = new HandshakeContext(privateKey);
+      }
+      
+      public NoiseProtocol(IEllipticCurveActions curveActions, IHkdf hkdf, 
+         ICipherFunction aeadConstruction, IKeyGenerator keyGenerator, IHashFunction hasher,
+         INoiseMessageTransformer messageTransformer, ILogger<NoiseProtocol> logger)
+      {
+         _curveActions = curveActions;
+         _hkdf = hkdf;
+         _aeadConstruction = aeadConstruction;
+         _keyGenerator = keyGenerator;
+         _hasher = hasher;
+         _messageTransformer = messageTransformer;
+         _logger = logger;
+      }
+      
+      public void InitHandShake(byte[] privateKey)
+      {
+         HandshakeContext = new HandshakeContext(privateKey);
+         
+         _logger.LogInformation("Initiating handshake hash");
+         
+         _hasher.Hash(LightningNetworkConfig.ProtocolNameByteArray(), HandshakeContext.ChainingKey);
+
+         _hasher.Hash(HandshakeContext.ChainingKey, LightningNetworkConfig.PrologueByteArray(), HandshakeContext.Hash);
+      }
+
+      // initiator act one
+      public void StartNewInitiatorHandshake(byte[] remotePublicKey, IBufferWriter<byte> output)
+      {
+         _logger.LogDebug("Initiator handshake: starting act one");
```

Comment:
Should we prefix all logs with some unique id to this instance?
It can be a short ver of the pubkey derived from the private key

Reply:
Added a simple session id so we can follow it if many handshakes happen in parallel, and remote public when known so we can trace issues.